File: problemskeleton/submissions/accepted/solution_correct.py
def count_sort_inverse(serial_numbers):
    if not serial_numbers:
        return []

    n = len(serial_numbers)
    max_serial = max(serial_numbers)

    # freq of each
    freqArr = [0] * (max_serial + 1)

    # get freq of each element
    for serial in serial_numbers:
        freqArr[serial] += 1

    # (special part for this question) compute sum backwards
    # freqArr[x] = nb of elements >= x
    for i in range(max_serial - 1, -1, -1):
        freqArr[i] += freqArr[i + 1]
    
    
    # build sorted array
    sorted_serials = [0] * n
    
    # stable: iterate backwards
    for i in range(n - 1, -1, -1):
        serial = serial_numbers[i]
        sorted_serials[freqArr[serial] - 1] = serial
        freqArr[serial] -= 1

    return sorted_serials#[::-1]
    
def main():
    nb_pieces_received = int(input())
    list_serial_nb = list(map(int, input().split()))
    
    sorted_serial_nb_pieces = count_sort_inverse(list_serial_nb)
    str_sorted_serial_nb_pieces = ' '.join(map(str, sorted_serial_nb_pieces))
    print(str_sorted_serial_nb_pieces)
    

if __name__ == "__main__":
    main()

File: problemskeleton/submissions/accepted/solution_correct.cpp
#include <iostream>
#include <vector>
using namespace std;

vector<int> count_sort_inverse(const vector<int>& serial_numbers) {
    int n = serial_numbers.size();
    if (n == 0) return {};

    // find max value
    int max_serial = 0;
    for (int s : serial_numbers) {
        if (s > max_serial) max_serial = s;
    }

    // create and initialize count array
    vector<int> countArr(max_serial + 1, 0);

    // get freq of each element
    for (int serial : serial_numbers) {
        countArr[serial]++;
    }

    //(special part for this question) compute sum backwards
    // freqArr[x] = nb of elements >= x
    for (int i = max_serial - 1; i >= 0; i--) {
        countArr[i] += countArr[i + 1];
    }

    // build sorted array
    vector<int> sorted_serials(n);

    // stable: iterate backwards
    for (int i = n - 1; i >= 0; i--) {
        int serial = serial_numbers[i];
        sorted_serials[countArr[serial] - 1] = serial;
        countArr[serial]--;
    }

    return sorted_serials;
}

int main() {
    int nb_pieces_received;
    cin >> nb_pieces_received;

    vector<int> list_serial_nb(nb_pieces_received);
    for (int i = 0; i < nb_pieces_received; i++) {
        cin >> list_serial_nb[i];
    }

    vector<int> sorted_serial_nb_pieces = count_sort_inverse(list_serial_nb);

    // print space-separated
    for (int i = 0; i < sorted_serial_nb_pieces.size(); i++) {
        if (i > 0) cout << " ";
        cout << sorted_serial_nb_pieces[i];
    }
    cout << endl;

    return 0;
}
